src/network/networkd-address.c

```diff
@@ -448,6 +448,127 @@ static int address_set_masquerade(Address *address, bool add) {
         return 0;
 }
 
+static int netlabel_op(bool add, const char *label, const Address *address) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *genl = NULL;
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        int r;
+
+        assert(address);
+        assert(address->link);
+        assert(address->link->network);
+
+        r = sd_genl_socket_open(&genl);
+        if (r < 0)
+                return r;
+
+        uint16_t command;
+        if (add)
+                command = NLBL_UNLABEL_C_STATICADD;
+        else
+                command = NLBL_UNLABEL_C_STATICREMOVE;
+
+        r = sd_genl_message_new(genl, NETLBL_NLTYPE_UNLABELED_NAME, command, &m);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_string(m, NLBL_UNLABEL_A_IFACE, address->link->ifname);
+        if (r < 0)
+                return r;
+
+        if (add) {
+                r = sd_netlink_message_append_string(m, NLBL_UNLABEL_A_SECCTX, label);
+                if (r < 0)
+                        return r;
+        }
+
+        union in_addr_union mask;
+
+        memset(&mask, 0xff, sizeof(mask));
+        r = in_addr_mask(address->family, &mask, address->prefixlen);
+        if (r < 0)
+                return r;
+
+        if (address->family == AF_INET) {
+                r = sd_netlink_message_append_in_addr(m, NLBL_UNLABEL_A_IPV4ADDR, &address->in_addr.in);
+                if (r < 0)
+                        return r;
+
+                r = sd_netlink_message_append_in_addr(m, NLBL_UNLABEL_A_IPV4MASK, &mask.in);
+        } else if (address->family == AF_INET6) {
+                r = sd_netlink_message_append_in6_addr(m, NLBL_UNLABEL_A_IPV6ADDR, &address->in_addr.in6);
+                if (r < 0)
+                        return r;
+
+                r = sd_netlink_message_append_in6_addr(m, NLBL_UNLABEL_A_IPV6MASK, &mask.in6);
+        }
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_send(genl, m, NULL);
```

Comment:
Please send the message asynchronously.

Reply:
I implemented asynchronous sending, but looking at the logs I'm not sure it's a good idea. I think at least for address configuration the operation should be synchronous, so that the label is used as soon as possible or at least before the address is configured. For deconfiguration doing label removal asynchronously probably would not be critical by itself, but then the removal operation could be mixed with a new address configuration event and the order of operations could get wrong. This is also made worse by a property of NetLabel operation for removal: it isn't symmetric with addition, the label can't be specified but all labels for the address are removed.

Reply:
order is one think, doing stuff synchronous is another. just make sure your calls here are enqueued before the netlink calls to add the addresses?

Reply:
Ok, kept the asynchronous version then.